crates/net/downloaders/src/headers/linear.rs

```diff
@@ -1,427 +1,857 @@
+}
+
+impl Future for HeadersRequestFuture {
+    type Output = HeadersRequestOutcome;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        let outcome = ready!(this.fut.poll_unpin(cx));
+        let request = this.request.take().unwrap();
+        Poll::Ready(HeadersRequestOutcome { request, outcome })
     }
 }
 
+/// The outcome of the [HeadersRequestFuture]
+struct HeadersRequestOutcome {
+    request: HeadersRequest,
+    outcome: PeerRequestResult<BlockHeaders>,
+}
+
+// === impl OrderedHeadersResponse ===
+
+impl HeadersRequestOutcome {
+    fn block_number(&self) -> u64 {
+        self.request.start.as_number().expect("is number")
+    }
+}
+
+/// Wrapper type to order responses
+struct OrderedHeadersResponse {
+    headers: Vec<Header>,
+    request: HeadersRequest,
+    peer_id: PeerId,
+}
+
+// === impl OrderedHeadersResponse ===
+
+impl OrderedHeadersResponse {
+    fn block_number(&self) -> u64 {
+        self.request.start.as_number().expect("is number")
+    }
+}
+
+impl PartialEq for OrderedHeadersResponse {
+    fn eq(&self, other: &Self) -> bool {
+        self.block_number() == other.block_number()
+    }
+}
+
+impl Eq for OrderedHeadersResponse {}
+
+impl PartialOrd for OrderedHeadersResponse {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for OrderedHeadersResponse {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.block_number().cmp(&other.block_number())
+    }
+}
+
+/// Type returned if a bad response was processed
+#[derive(Debug)]
+struct HeadersResponseError {
+    request: HeadersRequest,
+    peer_id: Option<PeerId>,
+    error: DownloadError,
+}
+
+/// The block to which we want to close the gap: (local head...sync target]
+#[derive(Debug)]
+struct SyncTargetBlock {
+    /// Block hash of the targeted block
+    hash: H256,
+    /// This is an `Option` because we don't know the block number at first
+    number: Option<u64>,
+}
+
 /// The builder for [LinearDownloader] with
 /// some default settings
 #[derive(Debug)]
 pub struct LinearDownloadBuilder {
     /// The batch size per one request
-    batch_size: u64,
-    /// The number of retries for downloading
-    request_retries: usize,
+    request_batch_size: u64,
+    /// Batch size for headers
+    stream_batch_size: usize,
+    /// Batch size for headers
+    min_concurrent_requests: usize,
+    /// Batch size for headers
+    max_concurrent_requests: usize,
+    /// How many responses to buffer
+    max_buffered_responses: usize,
 }
 
 impl Default for LinearDownloadBuilder {
     fn default() -> Self {
-        Self { batch_size: 100, request_retries: 5 }
+        Self {
+            request_batch_size: 1_000,
+            stream_batch_size: 10_000,
+            max_concurrent_requests: 150,
+            min_concurrent_requests: 5,
+            max_buffered_responses: 750,
+        }
```

Comment:
Would def like to play with this more, see how much we can stretch it. Could download 1m headers in 2m~, but seems like it could go faster with more peers etc.

Reply:
can probably be a bit more aggressive with these